### PR TITLE
As a user, my keyword details are scraped from Google search page

### LINKED ISSUE #14: As a user, my keyword details are scraped from Google search page
Scrap Google search page and store required details in the database.

## Acceptance Criteria

- Scrap first-page result of a Google search for each keyword
- Add new table report to store scraping results for each keyword

### PR DESCRIPTION
**Pull request by olivierobert**
Closes #14 

## What happened

Implement scraping from the Google search first page of results.

## Insight

The hardest part was to implementing testing properly hence the introduction of factories (and refactor). In addition, when it comes to testing the actual Google search I currently use mock both for `test` and `dev` environments (to avoid being blacklisted) using [this technique](http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/).

Known issue: using `Repo.Insert` in the Factory e.g. UserFactory.insert!(:user) does not behave the same as doing `Auth.create_user`. So I will need to solve that.

## Proof Of Work

*Jobs Table*

image

*Generated report*

image

### CODE REVIEW

File: lib/crawlexir/search/result_page.ex [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        },
+        organic_content: %{
+          count: 1,
+          url_list: ["https://some-links.com/"]
+        },
+        page_content: %{link_count: 100},
+        raw_html:  "<!doctype html><html> <> ..."
+      }
+  """
+  def new(body) do
```
Comment (rosle) [inline lib/crawlexir/search/result_page.ex]: I am thinking about maybe we can name it `parse`. When I saw new sometimes I'm thinking of a new object.

File: lib/crawlexir/search/keyword.ex [hunk 2]
```diff
@@ -3,19 +3,24 @@ defmodule Crawlexir.Search.Keyword do
...
   schema "keywords" do
     field :keyword, :string
 
     belongs_to :user, User
+    has_one :report, Report
 
     timestamps()
   end
 
+  @required_field ~w(keyword)a
```
Comment (rosle) [inline lib/crawlexir/search/keyword.ex]: I saw in the report, we put this on the top. Maybe we move this one too 👍

File: lib/crawlexir/search.ex [hunk 3]
```diff
@@ -76,61 +76,67 @@ defmodule Crawlexir.Search do
...
+      nil
 
   """
-  def change_keyword(%Keyword{} = keyword) do
-    Keyword.changeset(keyword, %{})
+  def get_keyword_report(keyword_id) do
+    Report
+    |> where(keyword_id: ^keyword_id)
+    |> preload(:keyword)
+    |> Repo.one()
```
Comment (rosle) [inline lib/crawlexir/search.ex:122]: Just wondering if this produce the same thing as `Repo.get_by(Keyword, keyword_id: keyword_id)`?

Comment (olivierobert) [reply lib/crawlexir/search.ex:122]: `Repo.get_by(Report, keyword_id: keyword_id)` would fetch the right record but not preload the associated keyword. Then I would need to do pipe preload:

```elixir
Repo.get_by(Report, keyword_id: keyword_id) |> Repo.preload(:keyword)
```

Personally, I find the current query piping nicer as it reads like `Report.where().preload()`. What do you think?

File: lib/crawlexir/search/keyword.ex (hunk 2 above)
Comment (olivierobert) [reply lib/crawlexir/search/keyword.ex]: Right, it's better indeed. Fixed in 4071a198995f83efbc52b1586f838bf48f342c50

File: lib/crawlexir/search/result_page.ex (hunk 1 above)
Comment (olivierobert) [reply lib/crawlexir/search/result_page.ex]: I am still unsure about that as well. I saw some lib using `new` but yes I also feel that `parse` would feel more functional programming oriented. Changed in f88d0052048318fa7db2dd3a4a16fd9af1c4190a

File: lib/crawlexir/search.ex (hunk 3 above)
Comment (rosle) [reply lib/crawlexir/search.ex:122]: Yep 👍

**Pull request merged**